src/resolve/resolved-dns-stream.c

```diff
@@ -441,6 +439,45 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
                         log_warning_errno(errno, "Couldn't restart TCP connection timeout, ignoring: %m");
         }
 
+        return 1;
+}
+
+
+static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *userdata) {
+        _cleanup_(dns_stream_unrefp) DnsStream *s = dns_stream_ref(userdata); /* Protect stream while we process it */
+        int r;
+
+        assert(s);
+
+        r = on_stream_io_impl(s, revents);
+        if (r <= 0)
+                return r;
+
+#if ENABLE_DNS_OVER_TLS
+        if (s->encrypted) {
+                uint32_t events;
+
+                r = sd_event_source_get_io_events(s->io_event_source, &events);
```

Comment:
Why this is called?? Why do not you reuse `revents` here?

Reply:
From `man 2 poll`, `revents` are the returned events (i.e. which events we received to process), `events` are the requested events (i.e. which events we could process next).

The point of checking whether `FLAGS_SET(events, EPOLLIN)` is to know whether the stream still wants to read more data. It would be wrong to return our "pseudo-`EPOLLIN` event " on buffered TLS data if the stream is not even expecting to receive further data available events.

Note that in this case `events` is just what was set by the last call to `dns_stream_update_io`. For example, looking at the condition on `dns_stream_update_io`, when we already have a full DNS packet stored on `s->read_packet`, we don't want to process any further data, even if it's available.

PS: BTW, something unrelated to this issue which I noticed when developing the test, is that `dns_stream_take_read_packet`doesn't call `dns_stream_update_io`. What this means is that unless you process the packets on the (supposedly optional) `on_packet` callback, taking the packet doesn't set the `EPOLLIN` flag again, so no further packets will be processed even if they are available... which seems wrong. But all callers use the `on_packet` callback, so this is just a theoretical issue for now.

Reply:
That makes sense for me. Thanks.

Reply:
> PS: BTW, something unrelated to this issue which I noticed when developing the test, is that `dns_stream_take_read_packet`doesn't call `dns_stream_update_io`. What this means is that unless you process the packets on the (supposedly optional) `on_packet` callback, taking the packet doesn't set the `EPOLLIN` flag again, so no further packets will be processed even if they are available... which seems wrong. But all callers use the `on_packet` callback, so this is just a theoretical issue for now.

I cannot catch the issue... There is `dns_stream_update_io()` after `on_packet()`. Is it not enough?

Reply:
Imagine I had written the loop of line 300 in `src/resolve/test-resolved-stream.c ` like this:
```c
        stream->on_packet = NULL;
        while (n_received_packets != 2) {
                assert_se(sd_event_run(event, EVENT_TIMEOUT_USEC) >= 1);
                DnsPacket *p = dns_stream_take_read_packet(stream);
                if (p)
                        received_packets[n_received_packets++] = p;
        }
```

IMO this ought to work, but it actually fails for both TCP and TLS, reading only a single packet, then timing out.

Why? Because when I call `sd_event_run`, `s->read_packet` is filled. When `dns_stream_update_io` is called after reading the packet in `on_stream_io_impl`, the `EPOLLIN` event flag is unset. Later I call `dns_stream_take_read_packet` so `s->read_packet` is free again, but no one sets the `EPOLLIN` event flag again, so no further packets are read and the test times out.

But all current usages call `dns_stream_take_read_packet` *inside* the `on_packet` callback rather than outside, so the `EPOLLIN` event flag is never unset, and everything works.

Reply:
Thank you for the explanation! Let's add a brief comment in `dns_stream_take_read_packet()`.

Reply:
BTW, while I try to understand your comment previously, I found a potential issue, and filed as #22266. If you'd like, PTAL.

Reply:
> Thank you for the explanation! Let's add a brief comment in `dns_stream_take_read_packet()`.

#22274.